crates/runtime/src/instance.rs

```diff
@@ -329,14 +412,29 @@ impl Instance {
 
     /// Return a reference to the vmctx used by compiled wasm code.
     #[inline]
-    pub fn vmctx(&self) -> &VMContext {
-        &self.vmctx
-    }
-
-    /// Return a raw pointer to the vmctx used by compiled wasm code.
-    #[inline]
-    pub fn vmctx_ptr(&self) -> *mut VMContext {
-        self.vmctx() as *const VMContext as *mut VMContext
+    pub fn vmctx(&self) -> *mut VMContext {
```

Comment:
Since this takes an immutable `&self`, should it return a `*const VMContext`, with a separate `vmctx_mut(&mut self)` or so if a `*mut VMContext` is needed?

Reply:
I won't pretend to be an expert on Stacked or Tree Borrows unfortunately, so all I can really say for sure is that this passes everything successfully there, so it's at least not a fundamental requirement. While I think it's best to model as carrying mutability it's not actually required due to the way the `vmctx` pointer is derived now (as it's derived from a pointer stored in `self` rather than purely from `self`). This means that the new `*_mut` methods from https://github.com/bytecodealliance/wasmtime/commit/0977952dcd9d482bff7c288868ccb52769b3a92e aren't actually required and we can technically go back to `vmctx_plus_offset(&self, u32) -> *mut T`, but I figured it's best to leave as-is for at least documentation purposes.

Reply:
Drive by comment: `*mut` and `*const` are, as far as I'm aware, treated as functionally the same in SB/TB, with their mutability determined by how they're derived, not their type. So, `&T -> *mut T -> &T` and `&mut T -> *const T -> &mut T` roundtrips are fine, but `&T -> *mut T -> &mut T` isn't.

This works here for the reasons alex mentioned - since it's derived from the original selfref pointer, it keeps its permissions as long as it isn't invalidated.